File: docs/algorithms/code/algorithms/canConstruct.ts
function canConstruct(ransomNote: string, magazine: string): boolean {
  if (magazine.length < ransomNote.length) {
    return false;
  }

  const map: Record<string, number> = {};

  for (let i = 0; i < magazine.length; i++) {
    if (map[magazine[i]]) {
      map[magazine[i]] += 1;
    } else {
      map[magazine[i]] = 1;
    }
  }

  for (let i = 0; i < ransomNote.length; i++) {
    if (!map[ransomNote[i]]) {
      return false;
    }
    map[ransomNote[i]] -= 1;
  }

  return true;
}

File: docs/algorithms/code/algorithms/canConstruct.go
func canConstruct(ransomNote string, magazine string) bool {
	hash := make(map[rune]int)

	if len(ransomNote) > len(magazine) {
		return false
	}

	for _, v := range magazine {
		if _, ok := hash[v]; !ok {
			hash[v] = 0
		}
		hash[v] = hash[v] + 1
	}

	for _, v := range ransomNote {
		if v, ok := hash[v]; !ok || v == 0 {
			return false
		}
		hash[v] = hash[v] - 1
	}

	return true
}